ALSpringBackend/src/test/java/com/ardaslegends/service/PathfinderTest.java

```diff
@@ -0,0 +1,193 @@
+                new Coordinate(1, 1, 1), new ArrayList<>(), new ArrayList<>(), new ArrayList<>(),
+                new ArrayList<>(), new ArrayList<>(), "none", "4", new HashSet<>());
+
+        // Set up relations
+        r1.getClaimedBy().add(faction_good);
+        r3.getClaimedBy().add(faction_bad);
+
+        player.setFaction(faction_good);
+        player.setRpChar(rpchar);
+
+        faction_good.getArmies().add(army);
+        faction_good.getPlayers().add(player);
+        faction_good.getRegions().add(r1);
+        faction_good.getRegions().add(r2);
+
+        faction_bad.getRegions().add(r3);
+
+        army.setBoundTo(rpchar);
+
+        r2.getClaimBuilds().add(claimbuild);
+
+        claimbuild.getSpecialBuildings().add(SpecialBuilding.HARBOUR);
+
+        // Set up region neighbours
+        r1.addNeighbour(r2);
+        r1.addNeighbour(r3);
+        r1.addNeighbour(r4);
+        r2.addNeighbour(r1);
+        r2.addNeighbour(r3);
+        r3.addNeighbour(r1);
+        r3.addNeighbour(r2);
+        r3.addNeighbour(r5);
+        r3.addNeighbour(r6);
+        r4.addNeighbour(r1);
+        r4.addNeighbour(r3);
+        r4.addNeighbour(r5);
+        r5.addNeighbour(r3);
+        r5.addNeighbour(r4);
+        r5.addNeighbour(r6);
+        r6.addNeighbour(r3);
+        r6.addNeighbour(r5);
+        rs1.addNeighbour(r2);
+        rs1.addNeighbour(r3);
+        rs2.addNeighbour(r3);
+        rs2.addNeighbour(r6);
+
+        List<Region> regionList = List.of(r1, r2, r3, r4, r5, r6, rs1, rs2);
+
+        mockRepository = mock(RegionRepository.class);
+
+        for (Region region : regionList) {
+            when(mockRepository.findById(region.getId())).thenReturn(Optional.of(region));
+        }
+
+        pathfinder = new Pathfinder(mockRepository);
+    }
+
+    @Test
+    void ensureNormalMoveSucceeds() {
+        Path path = pathfinder.findShortestWay("1", "2", player, false);
+        assertThat(path.getPath().size()).isEqualTo(2);
+        assertThat(path.getCost()).isEqualTo(RegionType.MOUNTAIN.getCost());
+
+        path = pathfinder.findShortestWay("1", "5", player, false);
+        assertThat(path.getPath().size()).isEqualTo(3);
+        assertThat(path.getCost()).isEqualTo(RegionType.LAND.getCost() + RegionType.HILL.getCost());
+
+        path = pathfinder.findShortestWay("1", "6", player, false);
+        assertThat(path.getPath().size()).isEqualTo(4);
+        assertThat(path.getCost()).isEqualTo(RegionType.LAND.getCost() * 2 + RegionType.HILL.getCost());
+    }
+
+    @Test
+    void ensureEmbarkingSucceeds() {
+        Region rs1 = mockRepository.findById("1.S").get();
+        Region rs2 = mockRepository.findById("2.S").get();
+        Region r2 = mockRepository.findById("2").get();
+        Region r3 = mockRepository.findById("3").get();
+        Region r6 = mockRepository.findById("6").get();
+        rs1.addNeighbour(rs2);
+        rs2.addNeighbour(rs1);
+        r2.addNeighbour(rs1);
+        r3.addNeighbour(rs1);
```

Comment:
Are the neighbours not already set in testData()? (same for the other tests)

Reply:
They are already set, but without sea regions (in terms of neighbours). So I add the sea regions as neighbours only in tests where I want the sea included, the rest only does land movement. Concerning your first comment, I only test army movement, but should add some for characters too indeed.